File: AoC2024_7.py

def fit_combination(key, index, factors, curr_sum):
    if index == len(factors):
        return curr_sum == key
    return fit_combination(key, index + 1, factors, curr_sum + factors[index]) or fit_combination(key, index + 1, factors, curr_sum * factors[index])

def fit_combination2(key, index, factors, curr_sum):
    if index == len(factors):
        return curr_sum == key
    
    if fit_combination2(key, index + 1, factors, curr_sum + factors[index]):
        return True
    if fit_combination2(key, index + 1, factors, curr_sum * factors[index]):
        return True
    if fit_combination2(key, index + 1, factors, int(str(curr_sum) + str(factors[index]))):
        return True 
    return False


def parse_file(filename):
    total = 0
    total2 = 0
    try:
        with open(filename, 'r') as file:
            for line in file:
                line = line.strip()
                if not line:
                    continue
                
                key, values = line.split(":")
                numbers = [int(x) for x in values.strip().split()]
                key = int(key)
                if fit_combination(key, 1, numbers, numbers[0]):
                    total += key
                if fit_combination2(key, 1, numbers, numbers[0]):
                    total2 += key
        print(f"Total: {total}")
        print(f"Total2: {total2}")
    except FileNotFoundError:
        print(f"File {filename} not found")
    except Exception as e:
        print(f"Error: {e}")
    finally:
        return total
                



print(parse_file("AoC2024_7.txt"))

File: AoC2024_7.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

type Equation struct {
	index    int
	target   uint64
	equation []uint64
}

func parseFile(filename string) []Equation {
	equations := make([]Equation, 0)
	file, err := os.Open(filename)
	if err != nil {
		log.Fatalf("Failed to open file: %v", err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	lineNumber := 0
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, ":")
		if len(parts) != 2 {
			log.Fatalf("Invalid line format at line %d: %s", lineNumber, line)
		}
		target, err := strconv.ParseInt(parts[0], 10, 64)
		if err != nil {
			log.Fatalf("Failed to parse target at line %d: %v", lineNumber, err)
		}
		equationParts := strings.Fields(parts[1])
		equation := make([]uint64, len(equationParts))
		for i, part := range equationParts {
			parsedInt, err := strconv.ParseInt(part, 10, 64)
			if err != nil {
				log.Fatalf("Failed to parse equation part at line %d: %v", lineNumber, err)
			}
			equation[i] = uint64(parsedInt)
			if err != nil {
				log.Fatalf("Failed to parse equation part at line %d: %v", lineNumber, err)
			}
		}
		equations = append(equations, Equation{
			index:    lineNumber,
			target:   uint64(target),
			equation: equation,
		})
		lineNumber++
	}
	return equations
}

func solveEquation(equation []uint64, index int, target uint64, currSum uint64) bool {
	if (index == len(equation)) {
		return currSum == target
	}
	return solveEquation(equation, index + 1, target, currSum + equation[index]) ||
		solveEquation(equation, index + 1, target, currSum * equation[index])
}

func main() {
	fmt.Println("Hello, playground.")
	equations := parseFile("AoC2024_7.txt")
	var total uint64 = 0
	for _, equation := range equations {
		if solveEquation(equation.equation, 0, equation.target, 0) {
			total += equation.target
		}
	} 
	fmt.Println("Total: ", total)
}
